docs/blueprints/selected-nodes-cache.md

```diff
@@ -0,0 +1,95 @@
+
+The cache should contain the following data:
+- already selected nodes and already selected **new** nodes (this may need to be 2 caches so we can select a smaller percentage of new nodes)
+- The only data that is sent to the uplink about the storage nodes is its address, typically being the IP and port. So the data that the cache should store is the storage node id, ip, port.
+
+The nodes table is the source of truth for the node data and should remain that way. The cache needs to be updated when it becomes stale. The cache should be updated when the following occur:
+
+A node should be removed from the cache when the following happen:
+- a node is disqualified, suspended, exited, out of space, hasn't been contacted recently
+
+A node should be updated in the cache when the following happen:
+- a node's address, port, or ip changes
+
+A node should be added to the cache when:
+- it is a new node on the network and is added to the nodes table for the first time (add to the unvetted selected nodes cache)
+
+A node should move from the unvetted to vetted cache when:
+- when it has completed enough audits and uptime checks to be vetted
+
+## Rationale
+
+Here are some other design that were considered:
+
+1) cache the entire nodes table
+
+This approach would allow more services to use this cache and benefit from the performance gains. However, it would also add a lot of complexity. Since the nodes table is the source, we need to update the cache anytime it becomes out of sync. Some queries might be able to be stale if it doesn't impact anything else, however it might get complex keeping track of what can and can't be stale. If the entire table is cached it will become outdated very often since so many services write to the table ( see background section above for a list). A fix to this could be to have the cache be the source of truth and frequently sync the cache to the nodes table. However this would be a big undertaking since it would require changing all the services that interact with the nodes table.
+
+2) Using Redis instead of in memory cache
+
+Using a remote cache like redis, might be preferred versus in memory since this cache will be used in each satellite api replica (which can currently scale up to 12).  The more api replicas we have, the more caches if we store in memory.
+
+Concerns about using a remote cache are:
+- the network latency retrieving data from the cache. Using the cache, selecting nodes should be very fast (estimating 0.1-0.5ms), so adding a network call out to redis will make this fast cache dramtically slower.
+
+Concerns about **not** using a remote cache are:
+- that this will cause more load on the database to keep all replica caches in sync (bandwidth and load),
+- every time we update the cache we may need to lock it and potentially block other work,
+- every replica will use up this much more space in memory which will increase over time and add up with each new replica.
```

Comment:
Another issue (sorry if I missed it and you discuss it elsewhere) is that it's not a shared cache between various satellite processes. (Maybe we don't horizontally scale the relevant processes? idk) For example, if multiple processes select a node for upload simultaneously that may result in the node not actually having enough disk space to complete the uploads.

Reply:
This will happen regardless since there are multiple satellites.

Reply:
If we use redis though they can share a cache. That isn't necessarily a reason to use it, but it has some advantages.

Reply:
So if we have 10 third-party satellites operated by different people we are going to share the cache between them?

Reply:
Do you mean if a single node is storing data from multiple satellites, they could be selected simultaneously for different satellites and thus not have enough space to store all the data.

We can't coordinate with satellites we don't operate, but we can coordinate different processes within a single satellite that we control with a shared cash.

It's not a panacea, but it does offer some benefits. That's the only point I'm trying to make.

Reply:
For the disk space problem, i proposed to the sno-growth-team to have an allocation of diskspace per satellite. Eg. you have 8TB, than its in best interested to divide the space up by the number of satellite and allow adjustion of it via the SNO Dashboard. That way an overcommitment of space is not possible and resolves the need for a shared cache

Reply:
The way we currently select nodes to upload data to (meaning without this cache), doesn't take into consideration if there are multiple upload requests occurring at the same time so that means this issue of a node not having enough disk space like Isaac mentions is currently a problem.  Additionally sharing one cache would not address that issue. It we wanted to prevent that problem I feel like we would need to know the last time the node was selected for an upload and if that upload was complete.